File: frontend/src/types/market.ts
export type IndicatorCard = {
  key: string;
  label: string;
  value: number | null;
  unit: string | null;
  reference_date: string | null;
};

export type CommodityCard = {
  asset: string;
  label: string;
  symbol: string | null;
  value: number | null;
  change_1d: number | null;
  change_7d: number | null;
  change_30d: number | null;
  change_90d: number | null;
  reference_date: string | null;
};

export type MarketOverview = {
  indicators: IndicatorCard[];
  commodities: CommodityCard[];
  data_as_of: string | null;
};

export type CurvePoint = {
  symbol: string;
  metric: string;
  value: number;
  reference_date: string;
  expiration_date: string | null;
  metadata: Record<string, unknown> | null;
};

export type CurveResponse = {
  asset: string;
  reference_date: string | null;
  points: CurvePoint[];
};

export type HistoryPoint = {
  metric: string;
  value: number;
  reference_date: string;
  metadata: Record<string, unknown> | null;
};

export type HistoryResponse = {
  symbol: string;
  points: HistoryPoint[];
};

export const RATE_CURVE_ASSETS = ["DI1", "DAP"] as const;
export type RateCurveAsset = (typeof RATE_CURVE_ASSETS)[number];

export const CURVE_WINDOWS = ["today", "7d", "30d", "90d"] as const;
export type CurveWindowLabel = (typeof CURVE_WINDOWS)[number];

export const CURVE_WINDOW_LABELS: Record<CurveWindowLabel, string> = {
  today: "Hoje",
  "7d": "7D",
  "30d": "30D",
  "90d": "90D",
};

export type RateCurvePoint = {
  symbol: string;
  expiration_date: string;
  reference_date: string;
  value: number;
};

export type RateCurveViewResponse = {
  asset: string;
  curves: Record<CurveWindowLabel, RateCurvePoint[]>;
};

export const TREASURY_ASSETS = ["treasury_ipca", "treasury_prefixado", "treasury_selic"] as const;
export type TreasuryAsset = (typeof TREASURY_ASSETS)[number];

export const TREASURY_LABELS: Record<TreasuryAsset, string> = {
  treasury_ipca: "Tesouro IPCA+",
  treasury_prefixado: "Tesouro Prefixado",
  treasury_selic: "Tesouro Selic",
};

export const TREASURY_IS_REAL_RATE: Record<TreasuryAsset, boolean> = {
  treasury_ipca: true,
  treasury_prefixado: false,
  treasury_selic: false,
};

export const TREASURY_COUPON_TYPE_LABELS: Record<string, string> = {
  zero: "Cupom zero",
  semestral: "Com juros semestrais",
};

export type TreasuryCurvePoint = {
  symbol: string;
  expiration_date: string;
  reference_date: string;
  bond_type: string | null;
  coupon_type: string | null;
  buy_rate: number | null;
  sell_rate: number | null;
};

export type TreasuryCurveViewResponse = {
  asset: string;
  coupon_types: string[];
  curves: Record<CurveWindowLabel, TreasuryCurvePoint[]>;
};

export const COMMODITY_ASSETS = ["BGI", "CCM", "ICF", "SJC"] as const;
export type CommodityAsset = (typeof COMMODITY_ASSETS)[number];

export const COMMODITY_HISTORY_PERIODS = ["30d", "90d", "6m", "1a"] as const;
export type CommodityHistoryPeriod = (typeof COMMODITY_HISTORY_PERIODS)[number];

export const COMMODITY_HISTORY_PERIOD_LABELS: Record<CommodityHistoryPeriod, string> = {
  "30d": "30D",
  "90d": "90D",
  "6m": "6M",
  "1a": "1A",
};

export type CommodityHistoryPoint = {
  reference_date: string;
  value: number;
};

export type CommodityHistoryApiResponse = {
  asset: string;
  symbol: string | null;
  points: CommodityHistoryPoint[];
};

File: backend/app/schemas/market.py
from datetime import date

from pydantic import BaseModel


class CurvePoint(BaseModel):
    symbol: str
    metric: str
    value: float
    reference_date: date
    expiration_date: date | None
    metadata: dict | None = None


class CurveResponse(BaseModel):
    asset: str
    reference_date: date | None
    points: list[CurvePoint]


class HistoryPoint(BaseModel):
    metric: str
    value: float
    reference_date: date
    metadata: dict | None = None


class HistoryResponse(BaseModel):
    symbol: str
    points: list[HistoryPoint]


class MacroObservation(BaseModel):
    value: float
    reference_date: date


class MacroSeriesResponse(BaseModel):
    slug: str
    points: list[MacroObservation]


class MacroResponse(BaseModel):
    series: list[MacroSeriesResponse]


class MetricPoint(BaseModel):
    category: str
    asset: str
    symbol: str | None
    metric: str
    value: float
    reference_date: date
    metadata: dict | None = None


class MetricsResponse(BaseModel):
    metrics: list[MetricPoint]


class IndicatorCard(BaseModel):
    key: str
    label: str
    value: float | None
    unit: str | None
    reference_date: date | None


class CommodityCard(BaseModel):
    asset: str
    label: str
    symbol: str | None
    value: float | None
    change_1d: float | None
    change_7d: float | None
    change_30d: float | None
    change_90d: float | None
    reference_date: date | None


class MarketOverviewResponse(BaseModel):
    indicators: list[IndicatorCard]
    commodities: list[CommodityCard]
    data_as_of: date | None


class RateCurvePoint(BaseModel):
    symbol: str
    expiration_date: date
    reference_date: date
    time_to_maturity_years: float
    value: float


class RateCurveViewResponse(BaseModel):
    asset: str
    curves: dict[str, list[RateCurvePoint]]


class TreasuryCurvePoint(BaseModel):
    symbol: str
    expiration_date: date
    reference_date: date
    time_to_maturity_years: float
    bond_type: str | None
    coupon_type: str | None
    buy_rate: float | None
    sell_rate: float | None


class TreasuryCurveViewResponse(BaseModel):
    asset: str
    coupon_types: list[str]
    curves: dict[str, list[TreasuryCurvePoint]]


class TreasuryBondHistoryPoint(BaseModel):
    reference_date: date
    buy_rate: float | None
    sell_rate: float | None
    buy_price: float | None


class TreasuryBondHistoryResponse(BaseModel):
    symbol: str
    asset: str | None
    bond_type: str | None
    coupon_type: str | None
    expiration_date: date | None
    rate_type: str | None
    current_buy_rate: float | None
    current_reference_date: date | None
    variation_bps: dict[str, float | None]
    points: list[TreasuryBondHistoryPoint]


class CommodityHistoryPoint(BaseModel):
    reference_date: date
    value: float


class CommodityHistoryResponse(BaseModel):
    asset: str
    symbol: str | None
    points: list[CommodityHistoryPoint]
